vault/logical_system_activation_flags.go

```diff
@@ -0,0 +1,123 @@
+var (
+	readActivationFlag = func(ctx context.Context, b *SystemBackend, req *logical.Request, fd *framework.FieldData) (*logical.Response, error) {
+		return b.readActivationFlag(ctx, req, fd)
+	}
+
+	writeActivationFlag = func(ctx context.Context, b *SystemBackend, req *logical.Request, fd *framework.FieldData, isActivate bool) (*logical.Response, error) {
+		return b.writeActivationFlagWrite(ctx, req, fd, isActivate)
+	}
+)
+
+func (b *SystemBackend) activationFlagsPaths() []*framework.Path {
+	return []*framework.Path{
+		{
+			Pattern: fmt.Sprintf("%s$", prefixActivationFlags),
+			DisplayAttrs: &framework.DisplayAttributes{
+				OperationVerb:   "read",
+				OperationSuffix: prefixActivationFlags,
+			},
+			Operations: map[logical.Operation]framework.OperationHandler{
+				logical.ReadOperation: &framework.PathOperation{
+					Callback: b.handleActivationFlagRead,
+					Summary:  summaryList,
+				},
+			},
+			HelpSynopsis:    helpSynopsis,
+			HelpDescription: helpDescription,
+		},
+	}
+}
+
+func (b *SystemBackend) handleActivationFlagRead(ctx context.Context, req *logical.Request, fd *framework.FieldData) (*logical.Response, error) {
+	return readActivationFlag(ctx, b, req, fd)
+}
+
+func (b *SystemBackend) handleActivationFlagsActivate(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	return writeActivationFlag(ctx, b, req, data, true)
+}
+
+func (b *SystemBackend) readActivationFlag(ctx context.Context, req *logical.Request, _ *framework.FieldData) (*logical.Response, error) {
+	activationFlags, err := b.Core.FeatureActivationFlags.Get(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return b.activationFlagsToResponse(activationFlags), nil
+}
+
+func (b *SystemBackend) writeActivationFlagWrite(ctx context.Context, req *logical.Request, _ *framework.FieldData, isActivate bool) (*logical.Response, error) {
+	// We need to manually parse out the feature_name from the path because we can't use FieldSchema parameters
+	// in the path to make generic endpoints. We need each activation-flag path to be a separate endpoint.
+	// Path starts out as activation-flags/<feature_name>/verb
+	// Removes activation-flags/ from the path
+	trimPrefix := strings.TrimPrefix(req.Path, prefixActivationFlags+"/")
+	// Removes /verb from the path
+	featureName := trimPrefix[:strings.LastIndex(trimPrefix, "/")]
+
+	err := b.Core.FeatureActivationFlags.Write(ctx, featureName, isActivate)
+	if err != nil {
+		return nil, fmt.Errorf("failed to write new activation flags: %w", err)
+	}
+
+	// We read back the value after writing it to storage so that we can try forcing a cache update right away.
+	// If this fails, it's still okay to proceed as the write has been successful and the cache will get updated
+	// at the time of an endpoint getting called. However, we can only return the one feature name we just activated
+	// in the response since the read to retrieve any others did not succeed.
```

Comment:
The behaviour of returning a different response but no error on a read error seems more confusing and prone to causing bugs to me than just to return an error in the rare occasion that the read fails even though the write just succeeded. I think it's mostly academic since the read is likely to be in LRU cache anyway if we only just wrote it OK which means that it is very unlikely to fail?

That said, if this is the behavior of the original implementation then there is an argument to not changing it to minimize changes in this PR.